File: 09/CCC '09 S4 - Shop and Ship.py
import math
infinity = math.inf
import sys
input = sys.stdin.readline
graph = [ 
]


n = int(input())
m = int(input())

# n, m = [int(x) for x in input().split()]

for x in range(n):
    graph.append([0] * n)

for x in range(m):
    a, b, c = [int(x) for x in input().split()] 
    if graph[a-1][b-1] != 0:
        if graph[a-1][b-1] > c:
            graph[a-1][b-1] = c 
            graph[b-1][a-1] = c
    else:
        graph[a-1][b-1] = c 
        graph[b-1][a-1] = c

pens = [False] * n
for x in range(int(input())):
    city, cost = [int(x) for x in input().split()] 
    pens[city-1] = cost 

src = int(input())-1
dist = [infinity] * n
dist[src] = 0


vis = [False] * n

cnt = 0


for x in range(n):
    # find unvisited node with min dist 
    min = infinity
    for y in range(n):
        if dist[y] < min and not vis[y]:
            min = dist[y]
            node = y 
    
    vis[node] = True

    for v in range(n):
        if graph[node][v] > 0 and not vis[v] and dist[v] > dist[node] + graph[node][v]:
            dist[v] = dist[node] + graph[node][v]

min = infinity 
for x in range(n):
    if dist[x] != infinity:
        if pens[x] != False:
            if dist[x] + pens[x] < min:
                min = dist[x] + pens[x]
print(min)

File: 09/CCC '09 S4 - Shop and Ship.cpp
#include <bits/stdc++.h>

using namespace std;

typedef pair<int, int> pi;

int n, t, p, src;

const int mm = 5e3+3;

vector<pi> graph[mm];

int pencil[mm], dist[mm], vis[mm];

int minNode() {
    int min_val = 1e9;
    int node = -1;

    for (int x = 1; x <= n; x++) {
        if (dist[x] < min_val && !vis[x]) {
            min_val = dist[x];
            node = x;
        }
    }
    return node;
}

int main() {
    cin.sync_with_stdio(0); cin.tie(0);    

    cin >> n >> t;

    for (int x = 0; x < t; x++) {
        int a, b, c; cin >> a >> b >> c;
        graph[a].push_back({b, c});
        graph[b].push_back({a, c});
    }

    for (int x = 1; x <= n; x++) {
        pencil[x] = 1e9;
        dist[x] = 1e9;
        vis[x] = false;
    }

    cin >> p;

    for (int x = 0; x < p; x++) {
        int a, b; cin >> a >> b;
        pencil[a] = b;
    }

    cin >> src;

    // run dijkstra

    dist[src] = 0;

    for (int x = 0; x < n; x++) {
        // find min node
        int node = minNode(); vis[node] = true;
        // update neighbours
        for (pi edge: graph[node]) {
            int v = edge.first, w = edge.second;
            dist[v] = min(dist[v], dist[node] + w);
        }
    }

    int ans = 1e9;

    for (int x = 1; x <= n; x++) ans = min(ans, dist[x] + pencil[x]);
    
    cout << ans << "\n";
}